src/wallet/wallet2.cpp

```diff
@@ -2253,78 +2074,90 @@ void wallet2::refresh(uint64_t start_height, uint64_t &blocks_fetched, bool &rec
 	// If stop() is called during fast refresh we don't need to continue
 	if(!m_run.load(std::memory_order_relaxed))
 		return;
-	pull_blocks(start_height, blocks_start_height, short_chain_history, blocks, o_indices);
-	// always reset start_height to 0 to force short_chain_ history to be used on
-	// subsequent pulls in this refresh.
-	start_height = 0;
-
-	while(m_run.load(std::memory_order_relaxed))
-	{
-		try
-		{
-			// pull the next set of blocks while we're processing the current one
-			uint64_t next_blocks_start_height;
-			std::list<cryptonote::block_complete_entry> next_blocks;
-			std::vector<cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::block_output_indices> next_o_indices;
-			bool error = false;
-			if(blocks.empty())
-			{
-				refreshed = false;
-				break;
-			}
-			tpool.submit(&waiter, [&] { pull_next_blocks(start_height, next_blocks_start_height, short_chain_history, blocks, next_blocks, next_o_indices, error); });
-
-			process_blocks(blocks_start_height, blocks, o_indices, added_blocks);
-			blocks_fetched += added_blocks;
-			waiter.wait();
-			if(blocks_start_height == next_blocks_start_height)
-			{
-				m_node_rpc_proxy.set_height(m_blockchain.size());
-				refreshed = true;
-				break;
-			}
 
-			// switch to the new blocks from the daemon
-			blocks_start_height = next_blocks_start_height;
-			blocks = next_blocks;
-			o_indices = next_o_indices;
+	//Download thread will likely be network or disk bound, so it should be in addition to max_conurrency
+	wallet_refresh_ctx refresh_ctx;
+	wallet_block_dl_ctx ctx(refresh_ctx);
+	size_t thd_max = std::min<size_t>(std::thread::hardware_concurrency(), 8);
```

Comment:
IMO it is better to query the maximum number of threads provided from the user (`tools::get_max_concurrency()`).